Don't create orphans when deleting compound

jonathangreen opened the pull request:
## JIRA Ticket:
https://jira.duraspace.org/browse/ISLANDORA-2294

This solution was discussed in the committers meeting yesterday.

## What does this Pull Request do?

Currently when deleting a compound we just remove relationships from children, so they are parentless and hard to detect. This PR changes that so that when deleting a compound we check if the children have parents. If they do we remove the relationships, otherwise we delete the children.

## What's new?

This updates compound deletes to work more like books and collections. It runs a batch that deletes the items or removes the relationships if they have other parents.

## How should this be tested?

I think the most important part of this PR is making sure that the other parent detection works correctly. We don't want to delete objects that have other parents. 

Some test cases: 
- Add a compound
- Batch import objects into compound (won't have other parents) 
- Add a book
- Add book pages
- Add a book page to the compound
- Add a item to the collection
- Add the item in the collection to the compound
- Add a newspaper, add an issue, add a page
- Add the issue and page to the compound
- Delete the compound and make sure that the correct objects are deleted and for the others the relationships are removed.

## Interested parties
@DiegoPino @Islandora/7-x-1-x-committers

jonathangreen committed:
Don't create orphans when deleting compound

Instead of just removing relationships from children when deleting a
compound we check if the children have parents. If they do we remove
the relationships, otherwise we delete the children.

jordandukart commented on islandora_compound_object.module:
```diff
@@ -157,19 +157,45 @@ function islandora_compound_object_block_view($delta = '') {
- * Implements hook_CMODEL_PID_islandora_object_purged().
+ * Implements hook_form_FORM_ID_alter().
  */
-function islandora_compound_object_islandora_compoundcmodel_islandora_object_purged($pid) {
```
Do we still need a hook to handle cases of objects being purged not within the forms?

jonathangreen replied:
Books and collections both don't handle the case of purging without using the forms. Does compound need that functionality?

jonathangreen replied:
IMO its better to leave orphaned children with dangling references, then to remove all the references and make them parentless, as the orphan report can at least find them if the have dangling references, but its possible there is a use case I don't see here.

DiegoPino replied:
Regarding this, i feel the hook idea is good, but would be more effective to have something in islandora itself to deal with that need/any other module that leaves orphans behind. Personally this pull as it is now solves the main issue of removing/deleting via forms, which is a good start? @jordandukart did you had time to test this/want to review? Don't want to step on people's toes by pushing forward.

jordandukart replied:
I don't really have an opinion for or against this behavior was just pointing out a change that -could- effect someone's current assumptions.

willtp87 commented:
Dose this handle compounds that don't have the compound cmodel?

jonathangreen commented:
@willtp87, it does not and the old solution did not handle that case either. I don't see a good way to know its a compound to trigger this functionality if its a compound without a compound cmodel. I am open to suggestions though.

jonathangreen committed:
Slay the coding standards dragon.

willtp87 commented:
@jonathangreen  Sounds like a separate issue then. Could we rely on the particular relation being present? Not sure how safe it would be given the config option.  Could be togglable though. Advise opening a Jira ticket and not holding this PR.

jonathangreen committed:
Coding standards

Bow down to the tyrany of the coding standards gods! May they take
this scarificial commit and allow my code to pass freely into the
land of review.

jonathangreen commented:
Created: https://jira.duraspace.org/browse/ISLANDORA-2296

DiegoPino commented:
@jonathangreen out of scope maybe, but would it be too crazy to have an Orphaned objects collection? Like a private, islandora autogenerated collection that is not under islandora:root trunk/tree and we could move all possible orphans, from any other operation there? What i really don't like is a batch removing objects without me being able to decide if they are still usable. Like, lets say i want to move them to a new compound, but by mistake is start by removing the original parent first... not sure, maybe just paranoid.
That collection could be a nursery collection =) (so many choices for the thumbnail!) and generate a message for the admin every time an object is "promoted" to be part of it. Also: lWhat happens when permissions differ and deletion can not be executed? Like you own the parent and but not the children? 

Also, to bring into the discussion (wait.. is diego recalling one of his old modules? yes he is)

I used to have the following code  when dealing with Ontologies and a lot of unknown possible predicates connecting objects and then  people deleting the, and it worked pretty well [1]

Instead of having a per SP relationships removal routing, what about having globally on islandora a predicate removal  like this (only as reference, old code, was working but can be done better)
[1]
https://github.com/DiegoPino/islandora_solution_pack_ontologies-7.x-dev/blob/5688719ace82ae4cfd40bfdb42293c44272b8b1c/includes/batch.inc#L99-L151

Which uses the heavy simplistic query 
https://github.com/DiegoPino/islandora_solution_pack_ontologies-7.x-dev/blob/5688719ace82ae4cfd40bfdb42293c44272b8b1c/includes/utilities.inc#L116

And all was invoked on object purged  (
https://github.com/DiegoPino/islandora_solution_pack_ontologies-7.x-dev/blob/master/islandora_ontologies.module#L293-L297
See no CMODEL specific call, just the old fashioned idea that whatever was binding two objects together, if one died, that link needed to be removed.

Too funky 🍄 ?
(NO DCS in place in that time.. you can laugh all you want!)

jonathangreen commented:
@DiegoPino I like the idea of an Orphan Objects collection, but its not consistent with what we are doing in other solution packs and its a fairly big change from what I'm doing here already. I like the idea of keeping compound consistent with book and collection. Perhaps the idea of an Orphan Objects collection could be broken out into its own ticket for some future work? 

I like your idea about how to search for parents better then what I'm doing. I'd rather not have to go out to the triplestore if I can avoid it, since we have the objects and the `RELS-EXT`, but I'll expand the query to find more relationships.

jonathangreen committed:
Loosen parent check.

DiegoPino commented on includes/batch.inc:
```diff
@@ -0,0 +1,199 @@
+  foreach ($relationships as $relationship) {
+    $is_cmodel_relationship = $relationship['predicate']['namespace'] == FEDORA_MODEL_URI
+      && $relationship['predicate']['value'] == 'hasModel';
+    if (!$is_cmodel_relationship
```
good, good! Reminded me of a cases of removed cmodels and objects that stayed behind. Never orphans really, more like class-less, which is easier to fix

DiegoPino commented:
I requested some feedback a week ago on how to proceed since other had concerns. Any updates on this? Would love to have this merged before code freeze

DiegoPino commented:
Oh, just saw this
https://github.com/Islandora/islandora_solution_pack_compound/pull/146#discussion_r215705330
Thanks @jordandukart, will review then and merge if all ok!

DiegoPino commented:
Yep, slipped under the carpet (old here) merging in 24

DiegoPino commented:
@jonathangreen 👀 merged. Will deal with Jira tomorrow. Cool commit!
